src/Files.App/Filesystem/StorageItems/ZipStorageFile.cs

```diff
@@ -478,6 +490,43 @@ private IAsyncOperation<Stream> OpenZipFileAsync(FileAccessMode accessMode)
 			});
 		}
 
+		private async Task<TOut> RetryWithCredentials<TOut>(Func<Task<TOut>> func, Exception exception)
+		{
+			var handled = exception is SevenZipOpenFailedException szofex && szofex.Result is OperationResult.WrongPassword ||
+				exception is ExtractionFailedException efex && efex.Result is OperationResult.WrongPassword;
+			if (!handled || PasswordRequested is null)
+				throw exception;
+
+			var tcs = new TaskCompletionSource<StorageCredential>();
+			PasswordRequested?.Invoke(this, tcs);
+			Credentials = await tcs.Task;
+			return await func();
+		}
+		private async Task RetryWithCredentials(Func<Task> func, Exception exception)
+		{
+			var handled = exception is SevenZipOpenFailedException szofex && szofex.Result is OperationResult.WrongPassword ||
+				exception is ExtractionFailedException efex && efex.Result is OperationResult.WrongPassword;
+			if (!handled || PasswordRequested is null)
+				throw exception;
+
+			var tcs = new TaskCompletionSource<StorageCredential>();
+			PasswordRequested?.Invoke(this, tcs);
+			Credentials = await tcs.Task;
+			await func();
+		}
+
+		public ZipStorageFile InitFromParent(StorageCredential credentials, EventHandler<TaskCompletionSource<StorageCredential>> passwordRequested)
```

Comment:
Once again, I cannot check it myself but a few methods seems to be repeted between this file and FtpFolder for example. Could they be equivalent? If so, maybe they could be adressed with a `T RefactoredMethod<T>(args)`?

Reply:
Absolutely, but again I'm in doubt about where to put the merce methods.

Reply:
> Could they be equivalent?
*The InitFromParent methods are equivalent but for RetryWithCredentials needs to check different exceptions in Zip and Ftp classes.